### PR TITLE
feat(DST-1322): add `current` prop to Sidebar.Nav for auto active item

### PR DESCRIPTION
**Pull request by sebald**
## Summary

- Adds a `current` prop on `Sidebar.Nav` that resolves the active leaf automatically — pass `pathname` (string) for smart segment-aware matching, or a predicate `(href, key) => boolean` for full control.
- Removes the `active={pathname === '/...'}` boilerplate on every `Sidebar.Item`. Per-item `active` stays as a local override and stacks with the resolved match (no breaking changes).
- Branch auto-highlight (existing `findActiveBranch`) keeps working when a descendant matches.

```tsx
// Before
<Sidebar.Nav>
  <Sidebar.Item href="/overview" active={pathname === '/overview'}>Overview</Sidebar.Item>
  <Sidebar.Item href="/users" active={pathname.startsWith('/users')}>Users</Sidebar.Item>
</Sidebar.Nav>

// After
<Sidebar.Nav current={pathname}>
  <Sidebar.Item href="/overview">Overview</Sidebar.Item>
  <Sidebar.Item href="/users">Users</Sidebar.Item>
</Sidebar.Nav>
```

### String-mode matching rules

1. Normalize input — strip query (`?…`), hash (`#…`), trim trailing `/`.
2. Leaves only — branches don't participate (their derived href would double-match with the first-leaf descendant).
3. Exact match wins first.
4. Otherwise, longest segment-prefix match (`current` starts with `href + '/'`, segment boundary mandatory). Handles dynamic routes like `/users/abc-123` lighting up `/users`.
5. Root (`/`) is exact-only — never acts as a prefix.
6. Equal-length matches: first in document order wins.

Dev-mode `console.error` when two leaves share the same href (gated behind `process.env.NODE_ENV !== 'production'`).

### Decisions log

The full decision log (identifier strategy, function shape, prop name rationale, override semantics, branch handling) is in [DST-1322](https://reservix.atlassian.net/browse/DST-1322).

Linked: this is the follow-up to DST-1312.

## Test plan

- [x] `pnpm typecheck:only` — clean
- [x] `pnpm test:unit` — 1041/1041 passed (102/102 Sidebar)
- [x] 31 new tests cover normalization, exact/longest-prefix matching, root exact-only, segment boundary, query/hash/trailing slash, predicate mode, branch exclusion, dev-mode duplicate warning, override interaction, and branch auto-highlight via `current`
- [x] Stories (`Basic`, `Complex`, `WithActiveBranch`, `KeyboardNavigation`) updated to use `current` and still work
- [ ] Manual smoke test in Storybook

[DST-1322]: https://reservix.atlassian.net/browse/DST-1322?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: docs/content/components/navigation/sidebar/index.mdx [hunk 1]
```diff
@@ -87,6 +87,21 @@ The sidebar supports one level of nesting. More levels would force users to reme
 
 <ComponentDemo name="sidebar-drilldown" />
 
+### Active item
+
+The sidebar marks the user's current page so they always know where they are. Pass the current pathname (typically from your router) to the `current` prop on `Sidebar.Nav`, and the matching item lights up automatically:
```
Comment (sarahgm) [inline docs/content/components/navigation/sidebar/index.mdx:92]: would update this also in all the demos which has sidebar or? This is the perefered way instead of using `active` ?

File: packages/components/src/Sidebar/Sidebar.test.tsx [hunk 2]
```diff
@@ -1231,3 +1237,384 @@ test('keyboard shortcut Ctrl+B toggles sidebar', async () => {
...
+});
+
+test('normalizePath preserves root when input is empty after stripping', () => {
+  expect(normalizePath('?foo=bar')).toBe('/');
+});
+
+// --- resolveCurrent: string mode ---
+
+const navJsx = () => [
+  <SidebarItem key="home" id="home" href="/">
```
Comment (sarahgm) [inline packages/components/src/Sidebar/Sidebar.test.tsx]: not sure I get it, why can't you reuse the story here?

File: packages/components/src/Sidebar/Sidebar.test.tsx [hunk 3]
```diff
@@ -1231,3 +1237,384 @@ test('keyboard shortcut Ctrl+B toggles sidebar', async () => {
...
+
+  // Branch.href === "/users" via firstLeafHref. Should not trigger duplicate
+  // warning because branches are excluded from leaf duplicate-detection.
+  resolveCurrent(collection, '/users');
+
+  expect(spy).not.toHaveBeenCalled();
+  spy.mockRestore();
+});
+
+// --- applyCurrent ---
```
Comment (sarahgm) [inline packages/components/src/Sidebar/Sidebar.test.tsx]: maybe its better for readablitity to move the tests about the collection in an own test file?

File: docs/content/components/navigation/sidebar/index.mdx (hunk 1 above)
Comment (sebald) [reply docs/content/components/navigation/sidebar/index.mdx:92]: yes, wanted to do this afterwards maybe. together with the thing that triggered it #5300 

yes, it really bugged me during #5300 that this is not simpler.

File: packages/components/src/Sidebar/Sidebar.test.tsx [hunk 4]
```diff
@@ -1231,3 +1237,384 @@ test('keyboard shortcut Ctrl+B toggles sidebar', async () => {
...
+  expect(nested?.type === 'item' && nested.active).toBe(true);
+});
+
+// --- Sidebar.Nav `current` integration ---
+
+const renderSidebar = (
+  current?: string | ((href: string | undefined, key: string) => boolean)
+) =>
+  render(
+    <RouterProvider navigate={() => {}}>
```
Comment (sarahgm) [inline packages/components/src/Sidebar/Sidebar.test.tsx]: why can't you reuse here a story?

File: packages/components/src/Sidebar/Sidebar.test.tsx (hunk 2 above)
Comment (sebald) [reply packages/components/src/Sidebar/Sidebar.test.tsx]: this tests the collection helpers :)

File: packages/components/src/Sidebar/Sidebar.test.tsx [hunk 5]
```diff
@@ -1231,3 +1237,384 @@ test('keyboard shortcut Ctrl+B toggles sidebar', async () => {
...
+        </Sidebar.Provider>
+      </MarigoldProvider>
+    </RouterProvider>
+  );
+
+test('Sidebar.Nav current prop marks matching leaf as active', () => {
+  renderSidebar('/analytics');
+
+  const link = linkByText('Analytics');
+  expect(link).toHaveAttribute('aria-current', 'page');
```
Comment (sarahgm) [inline packages/components/src/Sidebar/Sidebar.test.tsx]: blank lines sometimes missing

File: packages/components/src/Sidebar/collection.ts [hunk 6]
```diff
@@ -105,6 +105,142 @@ const firstLeafHref = (nodes: SidebarNode[]): string | undefined => {
...
+ * Normalize a URL path for matching:
+ * - strip query (`?…`) and hash (`#…`)
+ * - strip trailing slash, except the root (`/`)
+ */
+export const normalizePath = (path: string): string => {
+  const noQuery = path.split('?')[0].split('#')[0];
+  if (noQuery === '' || noQuery === '/') return '/';
+  // Negative lookbehind anchors the match at the first `/` of a trailing run,
+  // preventing polynomial backtracking on inputs like `'/'.repeat(n) + 'x'`.
+  const trimmed = noQuery.replace(/(?<!\/)\/+$/, '');
```
Comment (sarahgm) [inline packages/components/src/Sidebar/collection.ts]: Drop the regex ((?<!\/)\/+$) and replace with a while loop that trims trailing slashes. Same behavior, no regex to reason about.

File: packages/components/src/Sidebar/collection.ts [hunk 7]
```diff
@@ -105,6 +105,142 @@ const firstLeafHref = (nodes: SidebarNode[]): string | undefined => {
...
+            `item should have a unique href.`
+        );
+      } else {
+        seen.set(href, leaf.key);
+      }
+    }
+  }
+
+  // 1. Exact match (first in document order wins).
+  for (const leaf of leaves) {
```
Comment (sarahgm) [inline packages/components/src/Sidebar/collection.ts]: A hint from claude: 

  These can be one loop:

 ``` let exact: string | null = null;
  let best: { key: string; len: number } | null = null;

  for (const leaf of leaves) {
    const href = normalizePath(leaf.href);

    if (href === target) {
      exact = leaf.key;
      break;
    }

    if (href !== '/' && target.startsWith(href + '/')) {
      if (!best || href.length > best.len) {
        best = { key: leaf.key, len: href.length };
      }
    }
  }

  return new Set(exact ? [exact] : best ? [best.key] : []);
```

File: packages/components/src/Sidebar/collection.ts (hunk 6 above)
Comment (sebald) [reply packages/components/src/Sidebar/collection.ts]: leaving this as a regex, the loop would basically reimplement the regex and is harder to maintain

File: packages/components/src/Sidebar/Sidebar.test.tsx (hunk 3 above)
Comment (sebald) [reply packages/components/src/Sidebar/Sidebar.test.tsx]: separated them :)

File: packages/components/src/Sidebar/Sidebar.test.tsx (hunk 2 above)
Comment (sebald) [reply packages/components/src/Sidebar/Sidebar.test.tsx]: These are pure unit tests that call `buildCollection()` / `resolveCurrent()` directly with raw `SidebarItem` JSX — stories render the full component tree so there's no way to extract just the children. The `navJsx` fixture is also specifically crafted with minimal href patterns to exercise edge cases (exact vs prefix match, root exclusion, segment boundaries) that the stories aren't designed for.

File: packages/components/src/Sidebar/Sidebar.test.tsx (hunk 4 above)
Comment (sebald) [reply packages/components/src/Sidebar/Sidebar.test.tsx]: updated

File: packages/components/src/Sidebar/Sidebar.test.tsx (hunk 5 above)
Comment (sebald) [reply packages/components/src/Sidebar/Sidebar.test.tsx]: updated

File: packages/components/src/Sidebar/Sidebar.test.tsx (hunk 2 above)
Comment (sebald) [reply packages/components/src/Sidebar/Sidebar.test.tsx]: moved everything into its own file

File: packages/components/src/Sidebar/collection.ts (hunk 7 above)
Comment (sebald) [reply packages/components/src/Sidebar/collection.ts]: implemented

File: docs/content/components/navigation/sidebar/index.mdx (hunk 1 above)
Comment (sebald) [reply docs/content/components/navigation/sidebar/index.mdx:92]: If its fine doing this later, please resolve comment :)

File: packages/components/src/Sidebar/Sidebar.test.tsx (hunk 4 above)
Comment (sarahgm) [reply packages/components/src/Sidebar/Sidebar.test.tsx]: same for other tests in the file.

Should I mention in future every line where I think it can be reuse?

Comment (sebald) [reply packages/components/src/Sidebar/Sidebar.test.tsx]: Went through every test in all three new/changed test files. Here's the breakdown:

### Tests already using stories ✅

Most of `Sidebar.test.tsx` already reuses stories — `Basic.Component`, `Complex.Component`, `WithActiveBranch.Component`, `DefaultCollapsed.Component`, `Controlled.Component`. This covers: keyboard navigation (arrow, Home, End, wrapping), roving tabindex, Ctrl+B / Cmd+B shortcuts, collapse/expand, cookie persistence, mobile sheet, sub-panel open/back, branch switching, data-state, inert, separator, focus management, item id/key, and the `current` auto-opens-branch test.

### Tests that can't use stories (with reasons)

**`collection.test.tsx`** (entire file) — Pure unit tests that call `buildCollection()`, `resolveCurrent()`, `normalizePath()`, `applyCurrent()`, `findActiveBranch()` directly with raw JSX. Stories render the full component tree; there's no way to extract just the children to pass to these functions. The test fixtures are also deliberately minimal to exercise specific edge cases (exact vs prefix match, root exclusion, segment boundaries, duplicate href warnings) that the stories aren't designed for.

**`useSidebarNav.test.tsx`** (entire file) — Pure utility/hook tests: `panelPosition` is a plain function, `useRovingItem` tests an error thrown outside provider context, `useLastDistinctValue` tests hook state transitions via `renderHook`. None of these involve rendering a Sidebar component.

**`Sidebar.test.tsx` inline fixtures** — Each of these needs something no story provides:

| Test | Why it can't use a story |
|---|---|
| `onPress callback fires on item click` | Needs to pass a `vi.fn()` mock as `onPress` and assert it was called |
| `branch item onPress fires` | Same — needs mock callback on a branch trigger |
| `nested branches derive href from deepest first leaf` | Tests 3-level nesting; no story has this structure |
| `branch without any leaf hrefs renders without href` | Tests edge case with hrefless children; no story covers this |
| `group label inside branch renders correctly` | Tests GroupLabel/Separator _inside_ a branch; stories only have them at root |
| `textValue auto-extracted from string children` | Needs a branch _without_ explicit `textValue` prop; all stories set it explicitly |
| `direct branch-to-branch switch via active prop` | Needs an external button that changes active state to trigger branch-to-branch transition |
| `panel transition focuses active item` | Needs an external button to activate a child mid-session (tests focus-on-reentry) |
| `panel transition falls back to back button` | Needs a branch where no child has `active`; stories always have active items |
| `marker components return null` | Calls `Sidebar.Item({})` etc. directly — no rendering at all |
| `current` integration tests (renderSidebar helper) | Need to control the `current` prop value directly per test; stories wire `current` to internal `useState` so there's no way to inject a specific value from outside |
| `explicit active prop wins as local override` | Tests conflict resolution between `current` and per-item `active` — not a scenario any story covers |
| `without current prop falls back to per-item active` | Tests the no-`current` fallback path; all stories now use `current` |

Comment (sarahgm) [reply packages/components/src/Sidebar/Sidebar.test.tsx]: collection and useSidebar test i understand, but you changed one test with a story in Sidebar.test file so I thought it might be possible to change the other ones in this file also :)

**Review by sarahgm — APPROVED** [review]

**Review by aromko — APPROVED** [review]
